### PR TITLE
CP-36095 configure Stunnel SNI for client connections

### PR DESCRIPTION
**Pull request by lindig**
Make stunnel clients verify the connection by asking for the server certificate. This only works with the public server certificate present on the client, which is currently not the default. Hence, this is behind a feature flag.

This is a draft because I still need to test it manually but want to give reviewers a heads up what is coming.

### CODE REVIEW

File: stunnel/stunnel.ml [hunk 1]
```diff
@@ -134,21 +135,25 @@ let config_file verify_cert extended_diagnosis host port =
...
+    ; "ciphers = " ^ Xcp_const.good_ciphersuites
+    ; "curve = secp384r1"
+    ]
   ; if verify_cert then
-      ["verify=2"
-      ; sprintf "checkHost=%s" host
+      [ ""
+      ; "# use SNI to request a specific cert. CAfile contains"
+      ; "# public certs of all hosts in the pool and must contain"
+      ; "# the cert of the server we connect to"
```
Comment (lippirk) [inline stunnel/stunnel.ml:146]: I think it's good to have these comments, but probably in the ocaml instead (this config is never actually written to disk)

**Review by lippirk — APPROVED** [review]

File: stunnel/stunnel.ml (hunk 1 above)
Comment (lindig) [reply stunnel/stunnel.ml:146]: I know that the configuration is transient but still think the place is correct. Either way, would change it if you prefer.

File: stunnel/stunnel.ml [hunk 2]
```diff
@@ -134,21 +135,25 @@ let config_file verify_cert extended_diagnosis host port =
     ]
   ; if is_fips then ["fips=yes"] else ["fips=no"]
   ; if extended_diagnosis then ["debug=authpriv.7"] else ["debug=authpriv.5"]
+  ; [ "sslVersion = TLSv1.2"
+    ; "ciphers = " ^ Xcp_const.good_ciphersuites
+    ; "curve = secp384r1"
+    ]
```
Comment (psafont) [inline stunnel/stunnel.ml:141]: Although I find defining ssl version and ciphers here strange, it makes sense. I can't figure why is the curve parameter needed here.

**Review by psafont — APPROVED** [review]

**Comment by lindig** [comment]
I have tested this manually:

* Created `xapi-stunnel-ca-bundle.pem` in `/etc/stunnel` containing the public certs of a pool
* Created `/var/xapi/verify_certificates` to enable cert checking

It is safe to merge because the latter doesn't exist by default and hence the non=checking path is used.

**Pull request merged**